packages/store/src/weak-store.js

```diff
@@ -0,0 +1,40 @@
+// Copyright (C) 2019 Agoric, under Apache license 2.0
+
+// @ts-check
+
+import { assert, details, q } from '@agoric/assert';
+import './types';
+
+/**
+ * @template {Record<any, any>} K
+ * @template {any} V
+ * @param {string} [keyName='key']
+ * @returns {WeakStore<K, V>}
+ */
+export function makeWeakStore(keyName = 'key') {
+  const wm = new WeakMap();
+  const assertKeyDoesNotExist = key =>
+    assert(!wm.has(key), details`${q(keyName)} already registered: ${key}`);
+  const assertKeyExists = key =>
+    assert(wm.has(key), details`${q(keyName)} not found: ${key}`);
+  return harden({
+    has: key => wm.has(key),
+    init: (key, value) => {
```

Comment:
Was there consensus on the store API design before I joined?  Having `init` be distinct from `set` and having `delete` require the thing being deleted to first exist both seem like nods to purity that in my experience are ergonomic annoyances that people will just write wrappers around, resulting in a bunch of extra database accesses with no corresponding benefit.  Certainly the stores we use in the kernel don't have this.

Reply:
> Was there consensus on the store API design before I joined?

This was a feature of the store API from the beginning. It is the defining difference that caused us to coin the term "store" to distinguish it from JS's Maps. When did you join? That + the GitHub history can answer this question.

> my experience are ergonomic annoyances that people will just write wrappers around

For *defensive code* our experience is exactly the opposite. Tons of asserts that would have decorated each `get` and `set` magically disappeared by using stores rather than maps. I remember this being really striking in my early reviews of the Zoe implementation. Over and over again I would first think "we need to check here that...". Then I'd remember, "oh yeah, it is a store. It is already checking that."

Reply:
I wasn't involved the current `store` api, and did have some similar concerns. It was surprising how much code `init` in particular cleaned up for several use cases. 

The current API seems a little too limited to work for all our use cases, however. For example, we had some minor wallet bugs with the current semantics that had to be addressed by manually checking `has`. But it wasn't quite time to revisit the APIs until we had lots more examples, especially persistence. Now that we are doing persistence, we can revisit additional functionality. TBD whether it needs more operations, options in current operations, or a separate API for other use cases (e.g., a fancy query API). 

In both Xanadu and Midori collections, we had `get` vs `fetch` operations where `get` requires a value and `fetch` doesn't. My favorite missing operation for normal collection usage is an an efficient `provide` operation that returns a value if it's present or runs some code to produce it if it's not. In Smalltalk, that's the base operation that provides flexibility and reliability. In our `store`, that could be implemented with an optional second `ifAbsent` argument that is invoked if the object is not there.

Reply:
> or a separate API for other use cases (e.g., a fancy query API).

I hope to have other data structures exposed by the external store, such as a `makePriorityQueue(comparator)`, or somesuch.  I don't want to add more functionality to the `makeWeakStore()` API itself, unless it can be easily done for the in-memory version of the WeakStore, too.

> In our store, that could be implemented with an optional second ifAbsent argument that is invoked if the object is not there.

That sounds good!